docs: add sql review rules (#557)
bytebase/bytebase.com: content/docs/en/sql-review/review-rules/supported-rules.md

- TiDB
- PostgreSQL

<h3 id="naming.column.auto-increment">Auto-increment column naming convention</h3>

The unified naming convention is desired by developers. And the same applies to the database space. Bytebase provides this rule to unify the auto-increment column naming convention.

#### About convention format

> `Auto-increment Column Naming Convention` uses [regular expression](https://en.wikipedia.org/wiki/Regular_expression) format for naming pattern, and also limits the naming max length. The default maximum length is 64 characters.

Reviewer: nit: s/max/maximum

Always use full English words in articles, consistently at least.

##### Some typical format

| Name   | Regular Expression       |
| ------ | ------------------------ |
| id     | `^id$`                   |

#### How the rule works

> Bytebase checks that all auto-increment column names in DDL conform to the naming conventions.

Reviewer: ```suggestion
Bytebase checks all auto-increment column names in DDL conforming to the naming conventions.
```

- TiDB
- PostgreSQL

<h3 id="statement.disallow-commit">Disallow COMMIT</h3>

Disallow using commit in the issue.

#### How the rule works

Bytebase alert users if there are COMMIT statements.

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

<h3 id="statement.disallow-limit">Disallow LIMIT</h3>

Disallow LIMIT clause for INSERT, UPDATE and DELETE statements.

#### How the rule works

Specifically, Bytebase checks:

- `INSERT` statements
- `UPDATE` statements
- `DELETE` statements

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

<h3 id="statement.disallow-order-by">Disallow ORDER BY</h3>

Disallow ORDER BY clause for UPDATE and DELETE statements.

#### How the rule works

Specifically, Bytebase checks:

- `UPDATE` statements
- `DELETE` statements

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

<h3 id="statement.merge-alter-table">Merge ALTER TABLE</h3>

For more readability, it's better to not using multiply `ALTER TABLE` statements for same table.

#### How the rule works

Specifically, Bytebase checks:

> - `CREATE TABLE` statements

Author: Consider the following case:
```
CREATE TABLE table_t(...);
ALTER TABLE table_t ...;
ALTER TABLE table_t ...;
```
We need to merge them as one `CREATE TABLE` statements

Support for PostgreSQL is coming soon.

<h3 id="table.disallow-partition">Disallow partition table</h3>

#### How the rule works

Bytebase checks if SQLs creating the partition table.

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

<h3 id="table.comment">Table comment convention</h3>

Configure whether the table requires comments and the maximum comment length.

#### How the rule works

Bytebase checks the table comment convention.

Specifically, Bytebase checks:

> - `CREATE TABLE` statements

Reviewer 2: Why not ALTER TABLE?

- TiDB
- PostgreSQL

<h3 id="column.disallow-list">Column type disallow list</h3>

Set column type disallow list to ban column types.

#### How the rule works

> Bytebase checks if SQLs use the column type in disallow list.

Reviewer 2: ```suggestion
Bytebase checks if the SQL statement creates the column type in the disallow list.
```

- TiDB
- PostgreSQL

<h3 id="column.disallow-change-type">Disallow changing column type</h3>

> Changing column type may fail because the data cannot be converted. Bytebase provides this rule to alert you that a type changing has occurred.

Reviewer 2: ```suggestion
Changing column type may fail because the data cannot be converted. Bytebase provides this rule to alert you that the SQL statement would change the column type.
```

#### How the rule works

> Bytebase checks if SQLs change the column type.

Reviewer 2: ```suggestion
Bytebase checks if the SQL statement will change the column type.
```

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

<h3 id="column.set-default-for-not-null">Set DEFAULT value for NOT NULL columns</h3>

NOT NULL columns have no default value. It requires users to manually set default values for NOT NULL columns.

#### How the rule works

Bytebase checks if setting default values for NOT NULL columns.

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

<h3 id="column.disallow-change">Disallow ALTER TABLE CHANGE COLUMN statements</h3>

> CHANGE COLUMN is a MySQL extension to standard SQL. CHANGE COLUMN can change column definition and names, or both.
> Most of the time, you just want to change the one of two. So you need to use RENAME COLUMN and MODIFY COLUMN instead of CHANGE COLUMN to avoid unexpected modifications.

Reviewer 2: ```suggestion
CHANGE COLUMN is a MySQL extension to standard SQL. CHANGE COLUMN can change column definition and names, or both.
Most of the time, you just want to change one of two. So you need to use RENAME COLUMN and MODIFY COLUMN instead of CHANGE COLUMN to avoid unexpected modifications.
```

#### How the rule works

Bytebase checks if using `ALTER TABLE CHANGE COLUMN` statements.

#### Support database engine

- MySQL
- TiDB

<h3 id="column.disallow-changing-order">Disallow changing column order</h3>

Changing column order may cause performance issues. Users should be cautious about this.

#### How the rule works

Bytebase checks if changing column order.

Specifically, Bytebase checks:

- `ALTER TABLE` statements

#### Support database engine

- MySQL
- TiDB

<h3 id="column.auto-increment-must-integer">Use integer for auto-increment columns</h3>

> The auto-increment column must integer in MySQL.

Reviewer 2: ```suggestion
The auto-increment column must be integer.
```

- TiDB

Support for PostgreSQL is coming soon.

<h3 id="column.disallow-set-charset">Disallow set charset for columns</h3>

It's better to set the charset in the table or database.

#### How the rule works

Bytebase checks if setting charset for columns.

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements

#### Support database engine

- MySQL
- TiDB

<h3 id="column.auto-increment-must-unsigned">Set unsigned attribute on auto-increment columns</h3>

Setting unsigned attribute on auto-increment columns to avoid negative numbers.

#### How the rule works

Bytebase checks the unsigned attribute for auto-increment columns.

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements

#### Support database engine

- MySQL
- TiDB

<h3 id="column.comment">Column comment convention</h3>

Configure whether the column requires comments and the maximum comment length.

#### How the rule works

Bytebase checks the column comment.

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements

#### Support database engine

- MySQL
- TiDB

<h3 id="column.maximum-character-length">Maximum CHAR length</h3>

> The VARCHAR type is the variable-length type but the CHAR is not. When users need a long CHAR type, it's better to use VARCHAR instead of CHAR.

Reviewer 2: This reasoning is a bit unrelated. We should explain why we need to enforce CHAR length instead of why VARCHAR is better than CHAR

#### How the rule works

Bytebase checks the column default value.

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

## Index

<h3 id="index.no-duplicate-column">Disallow duplicate column in index keys</h3>

#### How the rule works

Bytebase checks if there exists duplicate column in index keys.

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements
- `CREATE INDEX` statements

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

<h3 id="index.key-number-limit">Limit the count of index keys</h3>

Limit the count of index keys in one index.

#### How the rule works

Bytebase checks the count of index keys in each index.

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements
- `CREATE INDEX` statements

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

<h3 id="index.pk-type-limit">Limit key type for primary keys</h3>

> Alert users if key type is not ING or BIGINT in primary keys.

Reviewer 2: ```suggestion
Alert users if key type is not INT or BIGINT in primary keys.
```

- TiDB

Support for PostgreSQL is coming soon.

## System

<h3 id="system.charset.allowlist">Charset allow list</h3>

#### How the rule works

> Bytebase checks if SQLs use the charset not in allow list.

Reviewer 2: ```suggestion
Bytebase checks if the SQL statement uses the charset outside of the allow list.
```

Specifically, Bytebase checks:

- `CREATE TABLE` statements
- `ALTER TABLE` statements

#### Support database engine

- MySQL
- TiDB

Support for PostgreSQL is coming soon.

<h3 id="system.collation.allowlist">Collation allow list</h3>

#### How the rule works

> Bytebase checks if SQLs use the collation not in allow list.

Reviewer 2: ```suggestion
Bytebase checks if the SQL statement uses the collation outside of the allow list.
```
